### PR TITLE
Update Draft + Deck schemas (Ready for review)

### PR DESCRIPTION
**Pull request by dekkerglen**
This update will update the draft + deck schemas to a single format (currently deck has 2 formats), and will accommodate multiplayer drafts in the future, as well as make it easier to reconstruct the draft pick by pick.

- [x] Update Schemas
- [x] Create one shot script for draft
- [x] Update all code that touches the draft object
- [x] Create one shot script for deck
- [x] Update all code that touches the deck object

### CODE REVIEW

File: models/deck.js [hunk 1]
```diff
@@ -20,10 +20,24 @@ Comment.add({
...
+  userid: String,
+  username: String,
+  pickorder: [],
+  name: String,
+  description: {
+    type: String,
+    default: 'No description available.',
+  },
+  cols: Number,
+  deck: [[]],
```
Comment (phulin) [inline models/deck.js]: I think we want to provide schemas for whatever is inside the array here.

File: models/deck.js [hunk 2]
```diff
@@ -32,35 +46,42 @@ let deckSchema = mongoose.Schema({
...
-  playerdeck: [[]],
-  playersideboard: [[]],
-  cols: Number,
-  username: {
+  comments: {
+    type: [Comment],
+    default: [],
+  },
+  draft: {
     type: String,
```
Comment (phulin) [inline models/deck.js]: Shouldn't this be `mongoose.Schema.Types.ObjectId`

File: models/draft.js [hunk 3]
```diff
@@ -1,16 +1,32 @@
...
+  name: String,
+  userid: String,
+  drafted: [[]], //organized draft picks
+  pickorder: [],
+  packbacklog: [[]],
+};
+
 // Cube schema
 let draftSchema = mongoose.Schema({
+  cube: String,
```
Comment (phulin) [inline models/draft.js]: same as above, `mongoose.Schema.Types.ObjectId`

File: models/draft.js [hunk 4]
```diff
@@ -1,16 +1,32 @@
 let mongoose = require('mongoose');
 
+//data for each seat, human or bot
+const Seat = {
+  bot: [], //null bot value means human player
+  name: String,
+  userid: String,
+  drafted: [[]], //organized draft picks
```
Comment (phulin) [inline models/draft.js]: can we add schemas to all of these arrays? cards should have at minimum a cardID

Comment (dekkerglen) [reply models/draft.js]: Definitely can do, and using strict schema rules here will mitigate database injection risks

**Comment by dekkerglen** [comment]
Functionally complete. Ready for more vigorous reviews. Will fix pending change requests.

**Comment by dekkerglen** [comment]
Oh yeah I forgot this would break a million unit tests, whoops

File: one_shot_scripts/update_deck_format.js [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+      };
+
+      deck.seats.push(playerSeat);
+
+      //add bots
+      for (let i = 1; i < deck.cards.length; i += 1) {
+        //need to build a deck with this pool...
+        const botdeck = await buildDeck(deck.cards[i]);
+        const bot = {
+          bot: deck.bots[i - 1],
```
Comment (RKelln) [inline one_shot_scripts/update_deck_format.js]: There is a bot for every card in the deck?

File: one_shot_scripts/update_draft_format.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    const bot = {
+      bot: draft.bots[i - 1],
+      name: 'Bot ' + i + ': ' + draft.bots[i - 1][0] + ', ' + draft.bots[i - 1][1],
+      pickorder: draft.picks[i],
+      drafted: [],
+      packbacklog: draft.packs[i] && draft.packs[i][0] ? [draft.packs[i][0]] : [],
+    };
+
+    //now we need to build picks from the pickorder ids
+    for (let j = 0; j < 16; j++) {
```
Comment (RKelln) [inline one_shot_scripts/update_draft_format.js]: What is this number from (15 picks?)? How does this work with custom drafts?

Comment (dekkerglen) [reply one_shot_scripts/update_draft_format.js]: 16 columns is the default display for decks, 8x2

File: src/components/DeckPreview.js [hunk 7]
```diff
@@ -8,7 +8,7 @@ const DeckPreview = ({ deck }) => {
   const maxLength = 35;
   const { date } = deck;
 
-  let { name } = deck;
+  let { name } = deck.seats[0];
```
Comment (RKelln) [inline src/components/DeckPreview.js]: Not using `.username` or `.name` here? Or is `name` not a good variable name for `deck.seats[0]`?

File: src/components/DeckbuilderNavbar.js [hunk 8]
```diff
@@ -118,6 +118,8 @@ const DeckbuilderNavbar = ({ deck, addBasics, name, description, className, ...p
     [saveForm],
   );
 
+  console.log(deck);
```
Comment (RKelln) [inline src/components/DeckbuilderNavbar.js]: Debugging

File: src/cube_overview.js [hunk 9]
```diff
@@ -197,7 +197,7 @@ class CubeOverview extends Component {
                   <CSRFForm
                     method="POST"
                     id="featuredForm"
-                    action={`/cube/${cube.isFeatured ? 'unfeature' : 'feature'}${cube._id}`}
+                    action={`/cube/${cube.isFeatured ? 'unfeature/' : 'feature/'}${cube._id}`}
```
Comment (RKelln) [inline src/cube_overview.js]: Slightly more DRY:
```
action={`/cube/${cube.isFeatured ? 'unfeature' : 'feature'}/${cube._id}`}
```

File: src/utils/Draft.js [hunk 10]
```diff
@@ -71,66 +78,123 @@ function botRating(botColors, card) {
...
+    sideboard,
+  };
+}
+
 function botPicks() {
   // make bots take one pick out of active packs
-  for (let botIndex = 1; botIndex < draft.packs.length; botIndex++) {
-    const botPack = draft.packs[botIndex][0];
-    const botColors = draft.bots[botIndex - 1];
+  for (let botIndex = 1; botIndex < draft.seats.length; botIndex++) {
```
Comment (RKelln) [inline src/utils/Draft.js]: If not subtracting 1 from index inside the loop, do you still want to start from 1?

File: src/utils/Draft.js [hunk 11]
```diff
@@ -71,66 +78,123 @@ function botRating(botColors, card) {
...
-    draft.packNumber += 1;
-    draft.pickNumber = 1;
-    // splice the first pack out
-    for (const drafter of draft.packs) {
-      drafter.splice(0, 1);
+  //check if pack is done
+  if (draft.seats.every((seat) => seat.packbacklog[0].length === 0)) {
+    //splice the first pack out
+    for (const seat of draft.seats) {
+      seat.packbacklog.splice(0, 1)[0];
```
Comment (RKelln) [inline src/utils/Draft.js]: Not setting anything = this? Shouldn't need `[0]` if you're just removing it?

File: src/utils/Draft.js [hunk 12]
```diff
@@ -139,27 +203,27 @@ async function pick(cardIndex) {
...
-        }
-      }
-    } else {
-      delete picks.details;
-    }
-  }
-  // save draft. if we fail, we fail
-  await csrfFetch(`/cube/api/draftpick/${draft.cube}`, {
+
+  const temp = JSON.parse(JSON.stringify(draft));
```
Comment (RKelln) [inline src/utils/Draft.js]: `temp` is debugging?

File: src/utils/draftutil.js [hunk 13]
```diff
@@ -1,11 +1,34 @@
...
-import { arrayShuffle } from './Util';
+'use strict';
+
+Object.defineProperty(exports, '__esModule', { value: true });
+
+var Util = require('./Util.js');
+require('./Card.js');
+var Filter = require('./Filter.js');
+
+function arrayShuffle(array) {
```
Comment (RKelln) [inline src/utils/draftutil.js]: Why is this here? Seems like you're still using `Util.arrayShuffle()` below?

File: one_shot_scripts/update_deck_format.js (hunk 5 above)
Comment (dekkerglen) [reply one_shot_scripts/update_deck_format.js]: deck.bots is where the old bot data is stored. it was an array of size equal to seats - 1, so 7 for an 8 person table. we need to pull this data into the seat object, which is why we do it like this

File: models/deck.js [hunk 14]
```diff
@@ -32,35 +46,42 @@ let deckSchema = mongoose.Schema({
...
+  username: {
+    type: String,
+    default: 'User',
+  },
+  cols: Number,
+  playerdeck: [[]],
+  playersideboard: [[]],
+  bots: [[]],
+  cards: [[]],
   newformat: {
```
[the remark this answers is not in the document]
Comment (dekkerglen) [reply models/deck.js]: Legacy schema, deprecated now

File: routes/cube_routes.js [hunk 15]
```diff
@@ -1,5 +1,6 @@
 const express = require('express');
 const fetch = require('node-fetch');
+``;
```
[the remark this answers is not in the document]
Comment (dekkerglen) [reply routes/cube_routes.js]: oh what's that doing there

File: src/utils/Draft.js (hunk 10 above)
Comment (dekkerglen) [reply src/utils/Draft.js]: 0 is the player seat, bot seats start at 1

File: models/deck.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (phulin) [reply models/deck.js]: FWIW I think we can remove those fields from the schema once the one-shot runs - they'll still be in the DB but inaccessible

File: src/utils/Draft.js (hunk 11 above)
Comment (dekkerglen) [reply src/utils/Draft.js]: correct

File: one_shot_scripts/update_deck_format.js [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+
+  const elos = await cubefn.getElo(cards.map((card) => card.details.name));
+  const nonlands = cards.filter((card) => !card.type_line.toLowerCase().includes('land'));
+  const lands = cards.filter((card) => card.type_line.toLowerCase().includes('land'));
+
+  sort_fn = function(a, b) {
+    if (bot) {
+      return botRating(b, bot, elos[b.details.name]) - botRating(a, bot, elos[a.details.name]);
+    } else {
+      return (elos[b.details.name] = elos[a.details.name]);
```
[the remark this answers is not in the document]
Comment (dekkerglen) [reply one_shot_scripts/update_deck_format.js]: Neither, it's supposed to subtraction

File: one_shot_scripts/update_deck_format.js [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+var cubefn = require('../serverjs/cubefn.js');
+
+const batch_size = 100;
+
+function arrayIsSubset(needles, haystack) {
+  return needles.every((x) => haystack.includes(x));
+}
+
+function botRating(botColors, card, rating) {
+  const colors = card.colors || card.details.color_identity;
```
Comment (ruler501) [inline one_shot_scripts/update_deck_format.js]: This handles colorless overrides incorrectly doesn't it since it's an empty array.

File: src/utils/Draft.js [hunk 18]
```diff
@@ -71,66 +78,123 @@ function botRating(botColors, card) {
...
+    for (let cardIndex = 0; cardIndex < pack.length; cardIndex++) {
+      if (draft.ratings && draft.ratings[pack[cardIndex].details.name]) {
         ratedPicks.push(cardIndex);
       } else {
         unratedPicks.push(cardIndex);
       }
     }
 
     ratedPicks.sort((x, y) => {
-      return botRating(botColors, botPack[y]) - botRating(botColors, botPack[x]);
```
Comment (ruler501) [inline src/utils/Draft.js]: So does this no longer depend on y?

File: one_shot_scripts/update_deck_format.js [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+const Deck = require('../models/deck');
+const Draft = require('../models/draft');
+const CardRating = require('../models/cardrating');
+const mongosecrets = require('../../cubecobrasecrets/mongodb');
+var carddb = require('../serverjs/cards.js');
+var cubefn = require('../serverjs/cubefn.js');
+
+const batch_size = 100;
+
+function arrayIsSubset(needles, haystack) {
```
[the remark this answers is not in the document]
Comment (dekkerglen) [reply one_shot_scripts/update_deck_format.js]: Not necessary for a one shot function

File: one_shot_scripts/update_deck_format.js (hunk 17 above)
Comment (dekkerglen) [reply one_shot_scripts/update_deck_format.js]: I think that's ok, we're just updating old decks anyway

Comment (dekkerglen) [reply one_shot_scripts/update_deck_format.js]: Bot decks are going to be far from ideal anyway

File: models/cardSchema.js [hunk 20]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
```
Comment (mawills) [inline models/cardSchema.js]: Excellent. These schema are sorely needed.
